### PR TITLE
Optimization staking

### PR DESCRIPTION
**Pull request by Jianzhuo**

### CODE REVIEW

File: contracts/Staking.sol [hunk 1]
```diff
@@ -201,27 +192,43 @@ contract Staking is IStaking, Initializable, OwnableUpgradeable, Constants {
         }
     }
 
-    function _checkDelegateLimitation(address _indexer, uint256 _amount) private view {
+    function _checkDelegateLimitation(address _indexer, uint256 _amount) external onlyStakingManager {
```
Comment (mzxyz) [inline contracts/Staking.sol]: Can remove `_` prefix for all these functions as they are not private any more

File: contracts/Staking.sol [hunk 2]
```diff
@@ -201,27 +192,43 @@ contract Staking is IStaking, Initializable, OwnableUpgradeable, Constants {
...
         require(
             delegation[_indexer][_indexer].valueAfter * indexerLeverageLimit >=
                 totalStakingAmount[_indexer].valueAfter + _amount,
             'S002'
         );
     }
 
+    function _registerIndexer(address _indexer) external onlyStakingManager {
```
Comment (mzxyz) [inline contracts/Staking.sol]: use `addIndexer` and `removeIndexer` as it only change the indexers related states but not include the full `indexerRegistry` features

File: contracts/Staking.sol [hunk 3]
```diff
@@ -471,12 +351,12 @@ contract Staking is IStaking, Initializable, OwnableUpgradeable, Constants {
...
         }
         
-        IERC20(settings.getSQToken()).safeTransfer(msg.sender, _amount);
+        IERC20(settings.getSQToken()).safeTransfer(settings.getDisputeManager(), _amount);
     }
 
     // -- Views --
 
-    function _isEmptyDelegation(address _source, address _indexer) internal view returns (bool) {
+    function _isEmptyDelegation(address _source, address _indexer) external view returns (bool) {
```
Comment (HZHAIX) [inline contracts/Staking.sol:359]: This function should be `internal` if only call inside `staking`

File: contracts/StakingManager.sol [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+import './interfaces/IIndexerRegistry.sol';
+
+contract StakingManager is IStakingManager, Initializable, OwnableUpgradeable {
+
+    ISettings private settings;
+
+    /**
+     * @dev Emitted when delegtor cancel unbond request.
+     */
+    event UnbondCancelled(address indexed source, address indexed indexer, uint256 amount, uint256 index);
```
Comment (mzxyz) [inline contracts/StakingManager.sol]: To make it consistent, keep this event inside staking contract

File: contracts/StakingManager.sol [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+    function slashIndexer(address _indexer, uint256 _amount) external {
+        require(msg.sender == settings.getDisputeManager(), 'G005');
+        Staking staking = Staking(settings.getStaking());
+        require(_amount <= staking.getSlashableAmount(_indexer), 'S010');
+        
+        staking._slashIndexer(_indexer, _amount);
+    }
+
+    function stakeCommission(address _indexer, uint256 _amount) external {
```
Comment (mzxyz) [inline contracts/StakingManager.sol]: The function to add commission fee to unbounding amount  can place inside `stakingContract`,

File: contracts/interfaces/IStakingManager.sol [hunk 6]
```diff
@@ -0,0 +1,26 @@
+// Copyright (C) 2020-2022 SubQuery Pte Ltd authors & contributors
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+pragma solidity 0.8.15;
+
+interface IStakingManager {
+    function stake(address _indexer, uint256 _amount) external;
+
+    function unstake(address _indexer, uint256 _amount) external;
```
Comment (mzxyz) [inline contracts/interfaces/IStakingManager.sol:9]: Review these interface, seems some of them are not used by other contracts

**Comment by mzxyz** [comment]
Awesome work! Just some small issues, also need to complete the function for adding commission fee to unbounding

**Comment by mzxyz** [comment]
🚀  LGTM.

**Review by mzxyz — APPROVED** [review]

**Pull request merged**